In [1]:
import pandas as pd

In [6]:
# Load into a DataFrame
worker_df = pd.read_csv('Worker.csv')
payitem_df = pd.read_csv('PayItem.csv')
job_df = pd.read_csv('Job.csv')
worker_itemized_amount_df = pd.read_csv('Worker_Itemized_Amount.csv')

In [7]:
worker_df.head()

WorkerID  WorkCode  HireDate EndDate City State
0       115         1  7/1/2015     NaN  NaN   NaN
1       210         4  7/1/2015     NaN  NaN   NaN
2       315         7  7/1/2015     NaN  NaN   NaN
3       420        10  7/1/2015     NaN  NaN   NaN
4       525        10  7/1/2015     NaN  NaN   NaN

In [9]:
df_merged_1 = pd.merge(worker_itemized_amount_df, worker_df, on='WorkerID', how='left')
df_merged_1

ID  WorkerID  WorkerInvoiceID  PayItemID  Quantity   UnitPay  \
0          1      4216         MH071515         14        33     0.575   
1          2      4216         MH071515         15         1    18.000   
2          3      4216         MH071515          1         1  1500.000   
3          4      4216         MH081515         14        41     0.575   
4          5      4216         MH081515         15         1    36.000   
...      ...       ...              ...        ...       ...       ...   
41312  41313      6954     DGreen0416-4         18         1   138.000   
41313  41314      7906     PJames0416-4         18         1   138.000   
41314  41315      9378    JCooper0416-4         18         1   138.000   
41315  41316     11170    MFisher0416-4         18         1   138.000   
41316  41317     11591  EHamilton0416-4         18         1   138.000   

      DatePerform  ItemizedAmount  WorkCode   HireDate EndDate          City  \
0       7/10/2015          18.975      13.0  7/10/2015     NaN      New York   
1       7/10/2015          18.000      13.0  7/10/2015     NaN      New York   
2       7/13/2015        1500.000      13.0  7/10/2015     NaN      New York   
3       7/27/2015          23.575      13.0  7/10/2015     NaN      New York   
4       7/27/2015          36.000      13.0  7/10/2015     NaN      New York   
...           ...             ...       ...        ...     ...           ...   
41312   4/22/2016         138.000      52.0  7/31/2015     NaN  Jacksonville   
41313   4/22/2016         138.000      52.0   8/5/2015     NaN      New York   
41314   4/22/2016         138.000      52.0  8/19/2015     NaN     Las Vegas   
41315   4/22/2016         138.000      52.0  9/14/2015     NaN       Seattle   
41316   4/22/2016         138.000      52.0  9/16/2015     NaN     Charlotte   

      State  
0        NY  
1        NY  
2        NY  
3        NY  
4        NY  
...     ...  
41312    FL  
41313    NY  
41314    NV  
41315    WA  
41316    NC  

[41317 rows x 13 columns]

In [11]:
df_merged_2 = pd.merge(df_merged_1, payitem_df, on='PayItemID', how='left')
df_merged_2

ID  WorkerID  WorkerInvoiceID  PayItemID  Quantity   UnitPay  \
0          1      4216         MH071515         14        33     0.575   
1          2      4216         MH071515         15         1    18.000   
2          3      4216         MH071515          1         1  1500.000   
3          4      4216         MH081515         14        41     0.575   
4          5      4216         MH081515         15         1    36.000   
...      ...       ...              ...        ...       ...       ...   
41312  41313      6954     DGreen0416-4         18         1   138.000   
41313  41314      7906     PJames0416-4         18         1   138.000   
41314  41315      9378    JCooper0416-4         18         1   138.000   
41315  41316     11170    MFisher0416-4         18         1   138.000   
41316  41317     11591  EHamilton0416-4         18         1   138.000   

      DatePerform  ItemizedAmount  WorkCode   HireDate EndDate          City  \
0       7/10/2015          18.975      13.0  7/10/2015     NaN      New York   
1       7/10/2015          18.000      13.0  7/10/2015     NaN      New York   
2       7/13/2015        1500.000      13.0  7/10/2015     NaN      New York   
3       7/27/2015          23.575      13.0  7/10/2015     NaN      New York   
4       7/27/2015          36.000      13.0  7/10/2015     NaN      New York   
...           ...             ...       ...        ...     ...           ...   
41312   4/22/2016         138.000      52.0  7/31/2015     NaN  Jacksonville   
41313   4/22/2016         138.000      52.0   8/5/2015     NaN      New York   
41314   4/22/2016         138.000      52.0  8/19/2015     NaN     Las Vegas   
41315   4/22/2016         138.000      52.0  9/14/2015     NaN       Seattle   
41316   4/22/2016         138.000      52.0  9/16/2015     NaN     Charlotte   

      State         PayItemName                  PayItemDescription  \
0        NY             mileage  Reimbursable mileage between sites   
1        NY             parking               Parking fees at sites   
2        NY      monthlyArticle                 Monthly news column   
3        NY             mileage  Reimbursable mileage between sites   
4        NY             parking               Parking fees at sites   
...     ...                 ...                                 ...   
41312    FL  training - lodging             Lodging during training   
41313    NY  training - lodging             Lodging during training   
41314    NV  training - lodging             Lodging during training   
41315    WA  training - lodging             Lodging during training   
41316    NC  training - lodging             Lodging during training   

       PayItemRate VariableRate  
0              NaN          NaN  
1              NaN          NaN  
2           1500.0          NaN  
3              NaN          NaN  
4              NaN          NaN  
...            ...          ...  
41312          NaN     Per Diem  
41313          NaN     Per Diem  
41314          NaN     Per Diem  
41315          NaN     Per Diem  
41316          NaN     Per Diem  

[41317 rows x 17 columns]

In [13]:
df_final_merged = pd.merge(df_merged_2, job_df, on='WorkCode', how='left')
df_final_merged

ID  WorkerID  WorkerInvoiceID  PayItemID  Quantity   UnitPay  \
0          1      4216         MH071515         14        33     0.575   
1          2      4216         MH071515         15         1    18.000   
2          3      4216         MH071515          1         1  1500.000   
3          4      4216         MH081515         14        41     0.575   
4          5      4216         MH081515         15         1    36.000   
...      ...       ...              ...        ...       ...       ...   
41312  41313      6954     DGreen0416-4         18         1   138.000   
41313  41314      7906     PJames0416-4         18         1   138.000   
41314  41315      9378    JCooper0416-4         18         1   138.000   
41315  41316     11170    MFisher0416-4         18         1   138.000   
41316  41317     11591  EHamilton0416-4         18         1   138.000   

      DatePerform  ItemizedAmount  WorkCode   HireDate EndDate          City  \
0       7/10/2015          18.975      13.0  7/10/2015     NaN      New York   
1       7/10/2015          18.000      13.0  7/10/2015     NaN      New York   
2       7/13/2015        1500.000      13.0  7/10/2015     NaN      New York   
3       7/27/2015          23.575      13.0  7/10/2015     NaN      New York   
4       7/27/2015          36.000      13.0  7/10/2015     NaN      New York   
...           ...             ...       ...        ...     ...           ...   
41312   4/22/2016         138.000      52.0  7/31/2015     NaN  Jacksonville   
41313   4/22/2016         138.000      52.0   8/5/2015     NaN      New York   
41314   4/22/2016         138.000      52.0  8/19/2015     NaN     Las Vegas   
41315   4/22/2016         138.000      52.0  9/14/2015     NaN       Seattle   
41316   4/22/2016         138.000      52.0  9/16/2015     NaN     Charlotte   

      State         PayItemName                  PayItemDescription  \
0        NY             mileage  Reimbursable mileage between sites   
1        NY             parking               Parking fees at sites   
2        NY      monthlyArticle                 Monthly news column   
3        NY             mileage  Reimbursable mileage between sites   
4        NY             parking               Parking fees at sites   
...     ...                 ...                                 ...   
41312    FL  training - lodging             Lodging during training   
41313    NY  training - lodging             Lodging during training   
41314    NV  training - lodging             Lodging during training   
41315    WA  training - lodging             Lodging during training   
41316    NC  training - lodging             Lodging during training   

       PayItemRate VariableRate                       Role Status  
0              NaN          NaN        Contributing Editor     IC  
1              NaN          NaN        Contributing Editor     IC  
2           1500.0          NaN        Contributing Editor     IC  
3              NaN          NaN        Contributing Editor     IC  
4              NaN          NaN        Contributing Editor     IC  
...            ...          ...                        ...    ...  
41312          NaN     Per Diem  Blog Special Contributors     IC  
41313          NaN     Per Diem  Blog Special Contributors     IC  
41314          NaN     Per Diem  Blog Special Contributors     IC  
41315          NaN     Per Diem  Blog Special Contributors     IC  
41316          NaN     Per Diem  Blog Special Contributors     IC  

[41317 rows x 19 columns]

In [14]:
filename = 'merged_dataframe.csv'
df_final_merged.to_csv(filename, index=False)

In [19]:
#Training Expenses
training_items = ['training - basePay', 'training - airFare', 'training - meals', 'training - lodging']
training_expenses = df_final_merged[df_final_merged['PayItemName'].isin(training_items) & (df_final_merged['Status'] == 'IC')]
total_training_expenses = training_expenses['ItemizedAmount'].sum()

#Travel Expenses
travel_items = ['mileage', 'parking', 'training - airFare']
travel_expenses = df_final_merged[df_final_merged['PayItemName'].isin(travel_items) & (df_final_merged['Status'] == 'IC')]
total_travel_expenses = travel_expenses['ItemizedAmount'].sum()

#Office Expenses
office_items = ['officeEquip - laptop', 'officeEquip - scanner', 'officeEquip - tablet']
office_expenses = df_final_merged[df_final_merged['PayItemName'].isin(office_items) & (df_final_merged['Status'] == 'IC')]
total_office_expenses = office_expenses['ItemizedAmount'].sum()

# Print results
total_training_expenses

102098.0

In [20]:
total_travel_expenses

71490.465

In [21]:
total_office_expenses

26556.0

In [23]:
lodging_during_training_count = df_final_merged[df_final_merged['PayItemDescription'] == "Lodging during training"].shape[0]
lodging_during_training_count

206

In [25]:
df_final_merged.shape[0]

41317

In [27]:
ic_lodging_training = df_final_merged[(df_final_merged['PayItemDescription'] == "Lodging during training") & (df_final_merged['Status'] == 'IC')]
unique_workerIDs_with_expense = ic_lodging_training['WorkerID'].unique()
unique_workerIDs_count = len(unique_workerIDs_with_expense)
total_unique_ic_workerIDs = df_final_merged[df_final_merged['Status'] == 'IC']['WorkerID'].nunique()

# Print
print(f"Number of unique IC WorkerIDs with 'Lodging during training' expenses: {unique_workerIDs_count}")
print(f"Total unique IC WorkerIDs in the dataset: {total_unique_ic_workerIDs}")

Number of unique IC WorkerIDs with 'Lodging during training' expenses: 54
Total unique IC WorkerIDs in the dataset: 74


This means that the workers are not classified correctly because the answer to the majority of the IRS criteria are Yes. For instance, the company is paying for lodging and training, the comapny is reimbersing travel expenses, and the company is paying office expenses.

In [28]:
from datetime import datetime

In [31]:
# Converting to datetime
df_final_merged['HireDate'] = pd.to_datetime(df_final_merged['HireDate'])
df_final_merged['EndDate'] = pd.to_datetime(df_final_merged['EndDate'])

# if NA make current date
current_date = pd.to_datetime('2016-08-01')
df_final_merged['EndDate'].fillna(current_date, inplace=True)


ic_df = df_final_merged[df_final_merged['Status'] == 'IC']

#length of employment
ic_df['LengthOfEmployment'] = (ic_df['EndDate'] - ic_df['HireDate']).dt.days

# average length of employment for ICs
average_length_of_employment_ic = ic_df['LengthOfEmployment'].mean()

average_length_of_employment_ic

/var/folders/x8/fqk_hz6j4x1f9s70rdmxj0d00000gn/T/ipykernel_12706/3427005936.py:13: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  ic_df['LengthOfEmployment'] = (ic_df['EndDate'] - ic_df['HireDate']).dt.days


318.0202584954981

This also makes us further confident that the companty is not properly classifying the employees because the average length of employment is 318 days which is a long term worker